MovingBlocks/Terasology
# Vive support

### Contains

Adds OpenVR support to Terasology.

When making my changes, my objective was to make this change completely invisible to people without VR headsets. There are many design decisions that will need to come downstream of this change, such as in-world overlays. This commit does not include any of that. If Terasology is launched with the `vrSupport` flag in `config.cfg` set equal to true, the world will be rendered in the headset (as well as on the screen). If not, everything acts exactly as it did before.

The designers of the game should draw their attention to the following aspects of this commit.
- I have removed any references to Oculus and TeraOVR.
- I renamed the flag `oculusVrSupport` to `vrSupport` for consistency.
- Since this uses OpenVR, it should work with both Oculus headsets and HTC Vive.
- OpenVR native libraries are included in `project_root/openvr_natives`. This directory must be present when the game is launched.
- My wrapper has been added directly to the Terasology source code / repository. I did this rather than keep it in a separate repository because the wrapper uses lwjgl, and compiling the wrapper to a jar would add a certain version dependency for lwjgl that could conflict with Terasology.
- A jar for `jopenvr` has replaced TeraOVR. This jar depends on JNA version 4.2.2, so as to follow Terasology. If the JNA dependency in Terasology ever changes, we will need to rebuild this jar. I felt this was cleaner than having a ton of machine-generated source code tagging along in the main repository, and the JNA version is unlikely to change in the future.
- Motion controller support is implemented through a listener. This listener is not used at the moment; this patch is already big enough, and that should be done separately and in a focused manner.
- As far as I know, this patch adds no work and actually subtracts some work to the directed acrylic graph effort. It's not programmed in an ideal way (My render frame buffers exist outside of their managing system), but it works as it is, mostly follows their design patterns (VR support is isolated to a single, new DAG node), and can be refactored later. The reason I left it like that is that it's clean/tidy that `openvrprovider` has no dependencies inside of Terasology and is completely self-contained.
- Ideally, down the road, we probably want to add some way that modules can add DAGs to the rendering pipeline and refactor VR support into an external module, but that is impossible with the DAG stuff in its current state. However, by merging this patch now, it's no more difficult to refactor later, and I can begin to write some of the support classes for VR interaction/gameplay.

Anyway, this has already gotten long enough. I'll add more comments later if I think of them.
### How to test
- If you have a VR headset, edit config.cfg such that `vrSupport` is set to true. Launch SteamVR. Launch the game. Edit your rendering settings to disable anything that could be incompatible with VR; for example "film grain" or "blur." Launch the game and put on your headset.
- If you have no VR headset, this patch should have no visible effect on your gameplay. If it does, I did something wrong and we should address it.
